src/llama-batch.cpp

```diff
@@ -323,6 +337,167 @@ struct llama_batch llama_batch_get_one(
     };
 }
 
+static struct llama_batch_ext * llama_batch_ext_init_impl(int32_t n_tokens_alloc, int32_t n_embd, int32_t n_seq_max) {
+    llama_batch_ext * batch = new llama_batch_ext{
+        /*n_tokens       =*/ 0,
+        /*max_tokens     =*/ n_tokens_alloc,
+        /*is_view        =*/ false,
+        /*tokens         =*/ nullptr,
+        /*embd           =*/ nullptr,
+        /*pos            =*/ nullptr,
+        /*n_seq_id       =*/ nullptr,
+        /*seq_id         =*/ nullptr,
+        /*logits         =*/ nullptr,
+    };
+
+    if (n_embd) {
+        batch->embd = (float *) malloc(sizeof(float) * n_tokens_alloc * n_embd);
+    } else {
+        batch->token = (llama_token *) malloc(sizeof(llama_token) * n_tokens_alloc);
+    }
+
+    batch->pos      = (llama_pos *)     malloc(sizeof(llama_pos)      * n_tokens_alloc);
+    batch->n_seq_id = (int32_t *)       malloc(sizeof(int32_t)        * n_tokens_alloc);
+    batch->seq_id   = (llama_seq_id **) malloc(sizeof(llama_seq_id *) * (n_tokens_alloc + 1));
+    for (int i = 0; i < n_tokens_alloc; ++i) {
+        batch->seq_id[i] = (llama_seq_id *) malloc(sizeof(llama_seq_id) * n_seq_max);
+    }
+    batch->seq_id[n_tokens_alloc] = nullptr;
+
+    batch->logits   = (int8_t *)        malloc(sizeof(int8_t)         * n_tokens_alloc);
+
+    return batch;
+}
+
+struct llama_batch_ext * llama_batch_ext_init(int32_t n_tokens_alloc, int32_t n_seq_max) {
+    return llama_batch_ext_init_impl(n_tokens_alloc, 0, n_seq_max);
+}
+
+struct llama_batch_ext * llama_batch_ext_init_from_embd(
+        const float * embd,
+             size_t   n_tokens,
+             size_t   n_embd,
+          llama_pos   pos0,
+       llama_seq_id   seq_id) {
+    struct llama_batch_ext * batch = llama_batch_ext_init_impl(n_tokens, n_embd, 1);
+    memcpy(batch->embd, embd, n_tokens * n_embd * sizeof(float));
+    for (size_t i = 0; i < n_tokens; i++) {
+        batch->pos     [i]    = pos0 + i;
+        batch->n_seq_id[i]    = 1;
+        batch->seq_id  [i][0] = seq_id;
+    }
+    return batch;
+}
+
+int32_t llama_batch_ext_set_pos(struct llama_batch_ext * batch, llama_pos * pos, size_t n_pos) {
+    if ((size_t) batch->n_tokens != n_pos) {
+        return -1;
+    }
+    memcpy(batch->pos, pos, n_pos * sizeof(llama_pos));
```

Comment:
I think this will copy only 1/4 of the positions for Qwen2VL.

Reply:
Edit: on closer look, seems like this is intentional. `image_embed` is divided into 4 batches hence only 1/4 positions being used. For now, I have no idea why.

Reply:
I think because on `master` the batch stores a pointer to a buffer with positions, it works correctly thanks to the `n_pos_per_token == 4`:

https://github.com/ggml-org/llama.cpp/blob/30f1db9936112bc04917df1a9520421ae5f3ffc1/src/llama-graph.cpp#L54-L60

But now this will likely do out-of-bounds read because we allocate a smaller buffer.

It's probably a good idea to figure this out, because it can change the API significantly. For example, `llama_batch_ext_init()` might need to accept `n_pos_per_token` argument (or a `llama_model` reference) so that it can allocate the appropriate `pos` array.

Reply:
Oh ok I didn't know that qwen2vl needs multiple pos per token.

Since qwen2vl is the only model supporting this, I think we should not add `n_pos_per_token` directly to `llama_batch_ext_init()`, this can make the API looks quite confusing.

Instead, we can simply allocate `batch.tokens` to `MAX_POS_PER_TOKEN * n_tokens` with `MAX_POS_PER_TOKEN` fixed to 4. In the future when the vision API is added, we could probably hide this position calculation into `libllama`. My idea looks like this:

```cpp
llama_mm_embd embd = llama_vision_get_output_embd(ctx);
llama_batch_ext * batch = llama_batch_ext_from_mm_embd(embd); // the mrope position is calculated here
```

Reply:
I proposed a fix in https://github.com/ggml-org/llama.cpp/pull/11875/commits/a9efdbbce56e95579b255d0052b289a76c1e7d9f

But this qwen2vl example seems to be broken even on `master`. I asked it to describe the windows xp bliss wallpaper and it gives wrong answers. Probably better to skip this and redo the whole thing on my vision refactoring.